File: node/src/telemetry/metrics.ts
// metrics.ts — counter/gauge/histogram aggregation for the telemetry ring.
// @public — см. ./README.md

import { MetricKind, type MetricPoint } from "../pb/servicebridge/v1/telemetry";

export { MetricKind };

export type Labels = Record<string, string>;

/**
 * MetricSink receives the points a drain produced. `TelemetryRing` satisfies it.
 * @internal
 */
export interface MetricSink {
	push(kind: "metrics", message: MetricPoint): void;
}

/**
 * MetricsHost owns a `MetricsAggregator`. `TelemetryRing` satisfies it; the
 * `make*` factories accept the host so callers keep passing the ring.
 * @internal
 */
export interface MetricsHost {
	readonly metrics: MetricsAggregator;
}

/** @public — см. ./README.md */
export interface Counter {
	inc(amount?: number): void;
}

/** @public — см. ./README.md */
export interface Gauge {
	set(value: number): void;
}

/** @public — см. ./README.md */
export interface Histogram {
	observe(value: number): void;
}

/**
 * Cumulative bucket entry as it is written to `MetricPoint.buckets_json`.
 * `count` is the number of observations `<= le`, so the `"+Inf"` entry carries
 * the total observation count. The overflow bound is the string `"+Inf"`
 * because JSON has no infinity literal.
 * @public — см. ./README.md
 */
export interface HistogramBucket {
	le: number | "+Inf";
	count: number;
}

/**
 * Default histogram bounds, in seconds — the Prometheus latency ladder.
 * @public — см. ./README.md
 */
export const DEFAULT_HISTOGRAM_BOUNDS: readonly number[] = [
	0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10,
];

const EMPTY_BUCKETS = Buffer.alloc(0);

interface SeriesBase {
	readonly name: string;
	readonly instanceId: string;
	readonly labels: Labels;
	readonly unit: string;
	dirty: boolean;
}

type CounterSeries = SeriesBase & {
	readonly kind: MetricKind.METRIC_KIND_COUNTER;
	sum: number;
};

type GaugeSeries = SeriesBase & {
	readonly kind: MetricKind.METRIC_KIND_GAUGE;
	value: number;
};

type HistogramSeries = SeriesBase & {
	readonly kind: MetricKind.METRIC_KIND_HISTOGRAM;
	readonly bounds: readonly number[];
	readonly counts: number[];
	sum: number;
};

type Series = CounterSeries | GaugeSeries | HistogramSeries;

// NUL separates key segments. A printable separator would let a label value
// forge one: with "|", `{a: "1|b"}` and `{a: "1", b: ""}` yield the same key and
// silently merge two series.
const SEP = "\u0000";

// Series identity. Label keys are sorted so {a:1,b:2} and {b:2,a:1} address the
// same state.
function seriesKey(
	kind: MetricKind,
	name: string,
	instanceId: string,
	sortedLabelKeys: string[],
	labels: Labels,
): string {
	let key = `${kind}${SEP}${name}${SEP}${instanceId}`;
	for (const k of sortedLabelKeys) key += `${SEP}${k}${SEP}${labels[k]}`;
	return key;
}

function normalizeLabels(labels: Labels): { keys: string[]; copy: Labels } {
	const keys = Object.keys(labels).sort();
	const copy: Labels = {};
	for (const k of keys) copy[k] = labels[k] as string;
	return { keys, copy };
}

function validateBounds(bounds: readonly number[]): readonly number[] {
	if (bounds.length === 0) {
		throw new Error("histogram bounds must not be empty");
	}
	for (let i = 0; i < bounds.length; i++) {
		const b = bounds[i] as number;
		if (!Number.isFinite(b)) {
			throw new Error(
				`histogram bound at index ${i} must be finite, got ${String(b)}`,
			);
		}
		if (i > 0 && b <= (bounds[i - 1] as number)) {
			throw new Error(
				`histogram bounds must be strictly ascending, got ${String(bounds[i - 1])} then ${String(b)}`,
			);
		}
	}
	return bounds;
}

function sameBounds(a: readonly number[], b: readonly number[]): boolean {
	if (a.length !== b.length) return false;
	for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
	return true;
}

function encodeBuckets(
	bounds: readonly number[],
	counts: readonly number[],
): Buffer {
	const out: HistogramBucket[] = [];
	let cumulative = 0;
	for (let i = 0; i < bounds.length; i++) {
		cumulative += counts[i] as number;
		out.push({ le: bounds[i] as number, count: cumulative });
	}
	cumulative += counts[bounds.length] as number;
	out.push({ le: "+Inf", count: cumulative });
	return Buffer.from(JSON.stringify(out));
}

/**
 * MetricsAggregator accumulates metric state per `(kind, name, instance, labels)`
 * series and emits exactly one `MetricPoint` per series per drain.
 *
 * Without it every `.inc()` was one ring item: a service at 1000 rps with one
 * counter increment per request produced 1000 points/s into a 16 KiB ring that
 * holds ~190 of them, so the ring silently oldest-dropped almost everything.
 *
 * Hot path (`inc`/`set`/`observe`) touches only the series object the handle
 * closes over — no map lookup, no allocation. The map is consulted once, when a
 * handle is created.
 *
 * @public — см. ./README.md
 */
export class MetricsAggregator {
	private readonly sink: MetricSink;
	private readonly series = new Map<string, Series>();

	constructor(sink: MetricSink) {
		this.sink = sink;
	}

	counter(instanceId: string, name: string, labels: Labels = {}): Counter {
		const s = this.resolve(
			MetricKind.METRIC_KIND_COUNTER,
			instanceId,
			name,
			labels,
			"1",
			() => ({ sum: 0 }),
		) as CounterSeries;
		return {
			inc(amount = 1): void {
				s.sum += amount;
				s.dirty = true;
			},
		};
	}

	gauge(instanceId: string, name: string, labels: Labels = {}): Gauge {
		const s = this.resolve(
			MetricKind.METRIC_KIND_GAUGE,
			instanceId,
			name,
			labels,
			"1",
			() => ({ value: 0 }),
		) as GaugeSeries;
		return {
			set(value: number): void {
				s.value = value;
				s.dirty = true;
			},
		};
	}

	histogram(
		instanceId: string,
		name: string,
		unit = "s",
		labels: Labels = {},
		bounds: readonly number[] = DEFAULT_HISTOGRAM_BOUNDS,
	): Histogram {
		const checked = validateBounds(bounds);
		const s = this.resolve(
			MetricKind.METRIC_KIND_HISTOGRAM,
			instanceId,
			name,
			labels,
			unit,
			() => ({
				bounds: checked,
				counts: new Array<number>(checked.length + 1).fill(0),
				sum: 0,
			}),
		) as HistogramSeries;
		if (!sameBounds(s.bounds, checked)) {
			throw new Error(
				`histogram "${name}" already registered with different bounds — one series has one bucket layout`,
			);
		}
		const bnds = s.bounds;
		const counts = s.counts;
		return {
			observe(value: number): void {
				let i = 0;
				while (i < bnds.length && value > (bnds[i] as number)) i++;
				counts[i] = (counts[i] as number) + 1;
				s.sum += value;
				s.dirty = true;
			},
		};
	}

	/**
	 * drain returns one point per series that changed since the previous drain and
	 * resets the accumulators. Counters and histograms reset to zero; a gauge keeps
	 * its last value but is not re-emitted until it is set again — re-sending an
	 * unchanged gauge every flush tick would put 4 points/s/series on the wire for
	 * no new information and re-create the ring overflow this class exists to fix.
	 * The runtime stores raw points, so the last emitted value stays readable.
	 */
	drain(nowMs: number = Date.now()): MetricPoint[] {
		const out: MetricPoint[] = [];
		for (const s of this.series.values()) {
			if (!s.dirty) continue;
			s.dirty = false;
			out.push(this.toPoint(s, nowMs));
		}
		return out;
	}

	/** flush drains and pushes every produced point into the sink. */
	flush(nowMs: number = Date.now()): number {
		const points = this.drain(nowMs);
		for (const p of points) this.sink.push("metrics", p);
		return points.length;
	}

	/** Number of distinct series currently tracked. @internal */
	get seriesCount(): number {
		return this.series.size;
	}

	private toPoint(s: Series, nowMs: number): MetricPoint {
		const base = {
			atUnixMs: nowMs,
			name: s.name,
			kind: s.kind,
			labels: s.labels,
			instanceId: s.instanceId,
			unit: s.unit,
		};
		if (s.kind === MetricKind.METRIC_KIND_COUNTER) {
			const value = s.sum;
			s.sum = 0;
			return { ...base, value, bucketsJson: EMPTY_BUCKETS };
		}
		if (s.kind === MetricKind.METRIC_KIND_GAUGE) {
			return { ...base, value: s.value, bucketsJson: EMPTY_BUCKETS };
		}
		const bucketsJson = encodeBuckets(s.bounds, s.counts);
		const value = s.sum;
		s.sum = 0;
		s.counts.fill(0);
		return { ...base, value, bucketsJson };
	}

	private resolve(
		kind: MetricKind,
		instanceId: string,
		name: string,
		labels: Labels,
		unit: string,
		init: () => object,
	): Series {
		const { keys, copy } = normalizeLabels(labels);
		const key = seriesKey(kind, name, instanceId, keys, copy);
		const existing = this.series.get(key);
		if (existing) {
			if (existing.unit !== unit) {
				throw new Error(
					`metric "${name}" already registered with unit "${existing.unit}", got "${unit}"`,
				);
			}
			return existing;
		}
		const created = {
			kind,
			name,
			instanceId,
			labels: copy,
			unit,
			dirty: false,
			...init(),
		} as Series;
		this.series.set(key, created);
		return created;
	}
}

// Counter is a monotonically increasing metric.
// @public — см. ./README.md
export function makeCounter(
	host: MetricsHost,
	instanceId: string,
	name: string,
	labels: Labels = {},
): Counter {
	return host.metrics.counter(instanceId, name, labels);
}

// Gauge can go up or down.
// @public — см. ./README.md
export function makeGauge(
	host: MetricsHost,
	instanceId: string,
	name: string,
	labels: Labels = {},
): Gauge {
	return host.metrics.gauge(instanceId, name, labels);
}

// Histogram records distribution observations into cumulative buckets.
// @public — см. ./README.md
export function makeHistogram(
	host: MetricsHost,
	instanceId: string,
	name: string,
	unit = "s",
	labels: Labels = {},
	bounds: readonly number[] = DEFAULT_HISTOGRAM_BOUNDS,
): Histogram {
	return host.metrics.histogram(instanceId, name, unit, labels, bounds);
}

File: go/internal/telemetry/metrics.go
package telemetry

import (
	"errors"
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
	"sync"

	pb "github.com/service-bridge/sdk/go/internal/pb/servicebridge/v1"
)

// ErrMetricConflict reports a metric resolved twice under one series key with a
// different shape — another unit, another bucket layout. One series has one
// shape; merging two shapes would make the stored points unreadable.
var ErrMetricConflict = errors.New("metric already registered with a different shape")

// ErrInvalidBounds reports histogram bounds that cannot describe buckets.
var ErrInvalidBounds = errors.New("invalid histogram bounds")

// Labels are the dimensions one metric series is keyed by.
type Labels map[string]string

// MetricSink receives the points one drain produced. *Ring satisfies it.
type MetricSink interface {
	PushMetric(*pb.MetricPoint)
}

// Units carried on the wire, in UCUM as the runtime stores them.
const (
	unitDimensionless = "1"
	unitSeconds       = "s"
	unitPercent       = "%"
	unitBytes         = "By"
)

// seriesSep joins the segments of a series key. It is NUL because a printable
// separator can be forged by a label value: under "|", {a: "1|b"} and
// {a: "1", b: ""} produce the same key and two unrelated series silently
// collapse into one.
const seriesSep = "\x00"

// DefaultHistogramBounds returns the latency ladder in seconds, as a fresh
// slice so a caller mutating it cannot corrupt other histograms.
func DefaultHistogramBounds() []float64 {
	return []float64{0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10}
}

// Metrics accumulates state per (kind, name, instance, labels) series and emits
// exactly one MetricPoint per changed series per drain.
//
// Aggregation is the whole point: one ring item per increment puts a service at
// 1000 rps far past what the metrics ring holds, and everything but the last
// few points is evicted before the transport ever sees it. The hot path
// (Inc/Set/Observe) touches only the series the handle closes over — the map is
// consulted once, when the handle is created, and no call allocates.
type Metrics struct {
	sink MetricSink

	mu     sync.Mutex
	series map[string]*series
	// order preserves creation order so a drain emits points deterministically;
	// Go map iteration would reshuffle them on every flush.
	order []*series
}

// NewMetrics binds an aggregator to the buffer its points are pushed into.
func NewMetrics(sink MetricSink) *Metrics {
	return &Metrics{sink: sink, series: make(map[string]*series)}
}

type series struct {
	kind       pb.MetricKind
	name       string
	instanceID string
	labels     map[string]string
	unit       string
	// bounds is immutable after creation, so Observe reads it without the lock.
	bounds []float64

	mu     sync.Mutex
	dirty  bool
	sum    float64
	value  float64
	counts []uint64
}

// Counter is a monotonically rising total.
type Counter struct{ s *series }

// Inc adds one.
func (c *Counter) Inc() { c.Add(1) }

// Add raises the counter by delta.
func (c *Counter) Add(delta float64) {
	c.s.mu.Lock()
	c.s.sum += delta
	c.s.dirty = true
	c.s.mu.Unlock()
}

// Gauge is a value that moves in both directions.
type Gauge struct{ s *series }

// Set replaces the current value.
func (g *Gauge) Set(value float64) {
	g.s.mu.Lock()
	g.s.value = value
	g.s.dirty = true
	g.s.mu.Unlock()
}

// Histogram distributes observations over cumulative buckets.
type Histogram struct{ s *series }

// Observe records one value.
func (h *Histogram) Observe(value float64) {
	s := h.s
	idx := len(s.bounds)
	for i, b := range s.bounds {
		if value <= b {
			idx = i
			break
		}
	}
	s.mu.Lock()
	s.counts[idx]++
	s.sum += value
	s.dirty = true
	s.mu.Unlock()
}

// Counter resolves the counter series and returns a handle on it. Two calls
// with the same instance, name and labels address the same series.
func (m *Metrics) Counter(instanceID, name string, labels Labels) (*Counter, error) {
	s, err := m.resolve(pb.MetricKind_METRIC_KIND_COUNTER, instanceID, name, unitDimensionless, labels, nil)
	if err != nil {
		return nil, fmt.Errorf("telemetry: counter %q: %w", name, err)
	}
	return &Counter{s: s}, nil
}

// Gauge resolves the gauge series and returns a handle on it.
func (m *Metrics) Gauge(instanceID, name, unit string, labels Labels) (*Gauge, error) {
	if unit == "" {
		unit = unitDimensionless
	}
	s, err := m.resolve(pb.MetricKind_METRIC_KIND_GAUGE, instanceID, name, unit, labels, nil)
	if err != nil {
		return nil, fmt.Errorf("telemetry: gauge %q: %w", name, err)
	}
	return &Gauge{s: s}, nil
}

// Histogram resolves the histogram series and returns a handle on it. An empty
// unit means seconds and nil bounds mean DefaultHistogramBounds. Bounds must be
// finite and strictly ascending, and must match what the series was created
// with.
func (m *Metrics) Histogram(instanceID, name, unit string, labels Labels, bounds []float64) (*Histogram, error) {
	if unit == "" {
		unit = unitSeconds
	}
	if bounds == nil {
		bounds = DefaultHistogramBounds()
	}
	if err := validateBounds(bounds); err != nil {
		return nil, fmt.Errorf("telemetry: histogram %q: %w", name, err)
	}
	s, err := m.resolve(pb.MetricKind_METRIC_KIND_HISTOGRAM, instanceID, name, unit, labels, bounds)
	if err != nil {
		return nil, fmt.Errorf("telemetry: histogram %q: %w", name, err)
	}
	return &Histogram{s: s}, nil
}

// Drain returns one point per series that changed since the previous drain and
// resets the accumulators. atUnixMs stamps every point.
func (m *Metrics) Drain(atUnixMs int64) []*pb.MetricPoint {
	m.mu.Lock()
	tracked := m.order
	m.mu.Unlock()

	out := make([]*pb.MetricPoint, 0, len(tracked))
	for _, s := range tracked {
		if p := s.point(atUnixMs); p != nil {
			out = append(out, p)
		}
	}
	return out
}

// Flush drains and pushes every produced point into the sink, returning how
// many points it emitted.
func (m *Metrics) Flush(atUnixMs int64) int {
	points := m.Drain(atUnixMs)
	for _, p := range points {
		m.sink.PushMetric(p)
	}
	return len(points)
}

// SeriesCount reports how many distinct series are tracked.
// @internal — см. ./README.md
func (m *Metrics) SeriesCount() int {
	m.mu.Lock()
	defer m.mu.Unlock()
	return len(m.series)
}

// point renders one series and resets what the emission consumed. A counter and
// a histogram start the next window empty; a gauge keeps its value but goes
// clean, so it is re-emitted only once it is set again. Re-sending an unchanged
// gauge on every flush tick would put a stream of points carrying no new
// information on the wire and re-create the ring overflow aggregation exists to
// prevent — the runtime stores raw points, so the last emitted value stays
// readable for as long as retention keeps it.
func (s *series) point(atUnixMs int64) *pb.MetricPoint {
	s.mu.Lock()
	if !s.dirty {
		s.mu.Unlock()
		return nil
	}
	s.dirty = false

	var (
		value   float64
		buckets []byte
	)
	switch s.kind {
	case pb.MetricKind_METRIC_KIND_COUNTER:
		value = s.sum
		s.sum = 0
	case pb.MetricKind_METRIC_KIND_GAUGE:
		value = s.value
	case pb.MetricKind_METRIC_KIND_HISTOGRAM:
		buckets = encodeBuckets(s.bounds, s.counts)
		value = s.sum
		s.sum = 0
		for i := range s.counts {
			s.counts[i] = 0
		}
	default:
		s.mu.Unlock()
		panic("telemetry: metric series with no kind")
	}
	s.mu.Unlock()

	return &pb.MetricPoint{
		AtUnixMs:    atUnixMs,
		Name:        s.name,
		Kind:        s.kind,
		Labels:      s.labels,
		InstanceId:  s.instanceID,
		Value:       value,
		Unit:        s.unit,
		BucketsJson: buckets,
	}
}

func (m *Metrics) resolve(
	kind pb.MetricKind,
	instanceID, name, unit string,
	labels Labels,
	bounds []float64,
) (*series, error) {
	keys, copied := normalizeLabels(labels)
	key := seriesKey(kind, name, instanceID, keys, copied)

	m.mu.Lock()
	defer m.mu.Unlock()

	if existing, ok := m.series[key]; ok {
		if existing.unit != unit {
			return nil, fmt.Errorf("unit %q, got %q: %w", existing.unit, unit, ErrMetricConflict)
		}
		if !sameBounds(existing.bounds, bounds) {
			return nil, fmt.Errorf("different bucket layout: %w", ErrMetricConflict)
		}
		return existing, nil
	}

	s := &series{
		kind:       kind,
		name:       name,
		instanceID: instanceID,
		labels:     copied,
		unit:       unit,
		bounds:     bounds,
	}
	if kind == pb.MetricKind_METRIC_KIND_HISTOGRAM {
		s.counts = make([]uint64, len(bounds)+1)
	}
	m.series[key] = s
	m.order = append(m.order, s)
	return s, nil
}

// normalizeLabels sorts the keys and copies the map, so the caller may reuse
// its own map and so {a,b} and {b,a} address one series.
func normalizeLabels(labels Labels) (keys []string, copied map[string]string) {
	if len(labels) == 0 {
		return nil, nil
	}
	keys = make([]string, 0, len(labels))
	for k := range labels {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	copied = make(map[string]string, len(labels))
	for _, k := range keys {
		copied[k] = labels[k]
	}
	return keys, copied
}

func seriesKey(kind pb.MetricKind, name, instanceID string, keys []string, labels map[string]string) string {
	var b strings.Builder
	b.WriteString(strconv.Itoa(int(kind)))
	b.WriteString(seriesSep)
	b.WriteString(name)
	b.WriteString(seriesSep)
	b.WriteString(instanceID)
	for _, k := range keys {
		b.WriteString(seriesSep)
		b.WriteString(k)
		b.WriteString(seriesSep)
		b.WriteString(labels[k])
	}
	return b.String()
}

func validateBounds(bounds []float64) error {
	if len(bounds) == 0 {
		return fmt.Errorf("no bounds: %w", ErrInvalidBounds)
	}
	for i, b := range bounds {
		if math.IsNaN(b) || math.IsInf(b, 0) {
			return fmt.Errorf("bound %d is not finite: %w", i, ErrInvalidBounds)
		}
		if i > 0 && b <= bounds[i-1] {
			return fmt.Errorf("bounds must ascend strictly, got %v then %v: %w", bounds[i-1], b, ErrInvalidBounds)
		}
	}
	return nil
}

func sameBounds(a, b []float64) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

// encodeBuckets renders the cumulative bucket table the runtime stores verbatim
// in JSONB: [{"le":<bound>,"count":<observations <= bound>}, ...] closed by an
// "+Inf" entry carrying the total. Without it a histogram is indistinguishable
// from a stream of gauges and no percentile can be computed from it. The bytes
// are built directly rather than marshalled: bounds are validated finite and
// counts are integers, so the output cannot fail to encode.
func encodeBuckets(bounds []float64, counts []uint64) []byte {
	buf := make([]byte, 0, len(bounds)*24+24)
	buf = append(buf, '[')
	var cumulative uint64
	for i, b := range bounds {
		cumulative += counts[i]
		if i > 0 {
			buf = append(buf, ',')
		}
		buf = append(buf, `{"le":`...)
		buf = strconv.AppendFloat(buf, b, 'g', -1, 64)
		buf = append(buf, `,"count":`...)
		buf = strconv.AppendUint(buf, cumulative, 10)
		buf = append(buf, '}')
	}
	cumulative += counts[len(bounds)]
	buf = append(buf, `,{"le":"+Inf","count":`...)
	buf = strconv.AppendUint(buf, cumulative, 10)
	buf = append(buf, '}', ']')
	return buf
}
